src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs

```diff
@@ -0,0 +1,362 @@
+                int endTagIndex,
+                out SyntaxNode rewritten,
+                out int startTagIndex)
+            {
+                rewritten = parent;
+                startTagIndex = -1;
+                var malformedTagCount = 0;
+
+                foreach (var tracker in _startTagTracker)
+                {
+                    if (tracker.TagName.Equals(tagName, StringComparison.OrdinalIgnoreCase))
+                    {
+                        break;
+                    }
+
+                    malformedTagCount++;
+                }
+
+                if (malformedTagCount != _startTagTracker.Count)
+                {
+                    parent = RewriteMalformedTags(parent, malformedTagCount);
+
+                    // One final rewrite, this is the rewrite that completes our target tag which is not malformed.
+                    var children = parent.ChildNodes();
+
+                    var startTagTracker = _startTagTracker.Pop();
+                    startTagIndex = startTagTracker.Index;
+                    var startTag = startTagTracker.TagBlock;
+                    
+                    // Get the nodes between the start and the end tag.
+                    var tagChildren = children.Skip(startTagIndex + 1).Take(endTagIndex - startTagIndex - 1).Cast<RazorSyntaxNode>();
+
+                    rewritten = RewriteNodeCore(parent, startTag, tagChildren, endTag);
+
+                    // We were able to recover
+                    return true;
+                }
+
+                // Could not recover tag. Aka we found an end tag without a corresponding start tag.
+                return false;
+            }
+
+            private SyntaxNode RewriteMalformedTags(SyntaxNode parent, int malformedTagCount)
+            {
+                for (var i = 0; i < malformedTagCount; i++)
+                {
+                    var children = parent.ChildNodes();
+                    var startTagTracker = _startTagTracker.Pop();
+                    var startTagIndex = startTagTracker.Index;
+                    var startTag = (MarkupTagBlockSyntax)children[startTagIndex];
+                    
+                    parent = RewriteNodeCore(parent, startTag, Enumerable.Empty<RazorSyntaxNode>(), endTag: null);
+                }
+
+                return parent;
+            }
+
+            private bool IsEndTag(MarkupTagBlockSyntax tagBlock)
+            {
+                var childSpan = (MarkupTextLiteralSyntax)tagBlock.Children.First();
+
+                // We grab the token that could be forward slash
+                var relevantToken = childSpan.LiteralTokens[childSpan.LiteralTokens.Count == 1 ? 0 : 1];
+
+                return relevantToken.Kind == SyntaxKind.ForwardSlash;
+            }
+
+            private bool IsVoidElement(MarkupTagBlockSyntax tagBlock)
+            {
+                return VoidElements.Contains(tagBlock.GetTagName(), StringComparer.OrdinalIgnoreCase);
```

Comment:
This is doing `Enumerable.Contains`. Did you mean to do `VoidElements.Contains(tagBlock.GetTagName())` since the hashset is already using an `OrdinalIgnoreCase` comparer?